### DISCUSSION TITLE
Sandbox creation fails due to failed size validation on air gap node

### DISCUSSION THREAD

**Discussion by ssmall** [k3s-io/k3s#8045] [Q&A] [answered]
**Environmental Info:**
K3s Version:  v1.27.3+k3s1 (fe9604ca)

Node(s) CPU architecture, OS, and Version: 
Linux _redacted_ 6.1.25-1rodete1-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.25-1rodete1 (2023-05-11) x86_64 GNU/Linux

Cluster Configuration:
3 (non-gapped) servers, 1 gapped agent

**Describe the bug:**
I followed the [air gapped installation instructions](https://docs.k3s.io/installation/airgap) to add an agent to my HA cluster. The agent starts successfully but no pods can be created successfully.

**Steps To Reproduce:**
follow the [air gapped installation instructions](https://docs.k3s.io/installation/airgap) to create a new agent, agent uses the following `/etc/racher/k3s/registries.yaml`:

```yaml
mirrors:
  docker.io:
    endpoint:
      - "https://us-west2-docker.pkg.dev/my-project-id/docker-proxy"
configs:
  us-west2-docker.pkg.dev:
    auth:
      username: redacted
      password: redacted
```

**Expected behavior:**

Pods in `kube-system` namespace start successfully on agent

**Actual behavior:**
Pods in `kube-system` stuck in `ContainerCreating`

**Additional context / logs:**

 I see errors in `journalctl -xe -u k3s-agent.service`:

```
kuberuntime_manager.go:1122] "CreatePodSandbox for pod failed" err="rpc error: code = FailedPrecondition desc = failed to get sandbox image \"rancher/mirrored-pause:3.6\": failed to pull image \"rancher/mirrored-pause:3.6\": failed to pull and unpack image \"docker.io/rancher/mirrored-pause:3.6\": failed commit on ref \"unknown-sha256:bed117227d0158a2e9bb99f725134398bf3c5f3ea70cc85b0a7061fc79c7b0f7\": \"unknown-sha256:bed117227d0158a2e9bb99f725134398bf3c5f3ea70cc85b0a7061fc79c7b0f7\" failed size validation: 621210 != 620442: failed precondition" pod="kube-system/svclb-traefik-be63bc17-5rfq9"
```

When I use `crictl` to pull the image from the private registry directly, the image is already up to date:

```
> sudo /usr/local/bin/k3s crictl pull us-west2-docker.pkg.dev/my-project-id/docker-proxy/rancher/mirrored-pause:3.6   
Image is up to date for sha256:6270bb605e12e581514ada5fd5b3216f727db55dc87d5889c790e4c760683fee
```

When I attempt to use `crictl` to pull the image by its docker.io path, I see the same size error:
```
> sudo /usr/local/bin/k3s crictl pull rancher/mirrored-pause:3.6                                                  
E0725 17:29:00.077520  569749 remote_image.go:167] "PullImage from image service failed" err="rpc error: code = FailedPrecondition desc = failed to pull and unpack image \"docker.io/rancher/mirrored-pause:3.6\": failed commit on ref \"unknown-sha256:4cf538ec773b05c10e5c512d0494145173604bf142dab736d073cce1b9b29fde\": \"unknown-sha256:4cf538ec773b05c10e5c512d0494145173604bf142dab736d073cce1b9b29fde\" failed size validation: 621050 != 620241: failed precondition" image="rancher/mirrored-pause:3.6"
FATA[0001] pulling image: rpc error: code = FailedPrecondition desc = failed to pull and unpack image "docker.io/rancher/mirrored-pause:3.6": failed commit on ref "unknown-sha256:4cf538ec773b05c10e5c512d0494145173604bf142dab736d073cce1b9b29fde": "unknown-sha256:4cf538ec773b05c10e5c512d0494145173604bf142dab736d073cce1b9b29fde" failed size validation: 621050 != 620241: failed precondition 

> sudo /usr/local/bin/k3s crictl pull docker.io/rancher/mirrored-pause:3.6                                                  
E0725 17:29:35.710905  569810 remote_image.go:167] "PullImage from image service failed" err="rpc error: code = FailedPrecondition desc = failed to pull and unpack image \"docker.io/rancher/mirrored-pause:3.6\": failed commit on ref \"unknown-sha256:c31d4f0683d3a062725688ba3058d9095f3982e52b1f3a1cec433398517a5bca\": \"unknown-sha256:c31d4f0683d3a062725688ba3058d9095f3982e52b1f3a1cec433398517a5bca\" failed size validation: 620958 != 620337: failed precondition" image="docker.io/rancher/mirrored-pause:3.6"
FATA[0001] pulling image: rpc error: code = FailedPrecondition desc = failed to pull and unpack image "docker.io/rancher/mirrored-pause:3.6": failed commit on ref "unknown-sha256:c31d4f0683d3a062725688ba3058d9095f3982e52b1f3a1cec433398517a5bca": "unknown-sha256:c31d4f0683d3a062725688ba3058d9095f3982e52b1f3a1cec433398517a5bca" failed size validation: 620958 != 620337: failed precondition 
```

**Comment by brandond** [comment, marked as answer]
```yaml
mirrors:
  docker.io:
    endpoint:
      - "https://us-west2-docker.pkg.dev/my-project-id/docker-proxy"
```

That is not now that works. The endpoint needs to be the base URI of the registry API on that host. Based on what you're pulling, the registry endpoint is just at the default location for that host: `https://us-west2-docker.pkg.dev/v2/`

What I think you're are attempting to do is **rewrite** the images when pulling from your mirror - you want to turn `rancher/mirrored-pause` into `my-project-id/docker-proxy/rancher/mirrored-pause` when pulling from the mirror. For that you want to reference this section of the documentation: https://docs.k3s.io/installation/private-registry#rewrites

**Reply by ssmall** [reply]
Thanks, that fixed it! Out of curiosity, it didn't seem to work when I tried to configure a full wildcard rewrite:

```yaml
rewrite:
      "^(.*)": "my-project-id/docker-proxy/$1"
```

with that config, k3s seemed to ignore the mirror configuration entirely and fell back to resolving images from `registry-1.docker.io/v2/`

when I updated the config to 

```yaml
rewrite:
      "^rancher/(.*)": "my-project-id/docker-proxy/rancher/$1"
```

everything works as expected, though.

Is this the expected behavior?

**Reply by brandond** [reply]
A full rewrite should work, assuming everything on docker-hub is available through that proxy. You don't need to anchor the regex though; did you try:
```yaml
"(.*)": "my-project-id/docker-proxy/$1"
```

What did you see that made you suspect it wasn't pulling from your endpoint?

**Reply by ssmall** [reply]
Getting rid of the anchor seems to have fixed it. What I was seeing previously when using `"^(.*)"` as the regex were log lines in `journalctl -xeu k3s-agent.service` like

```
failed to resolve reference \"docker.io/rancher/mirrored-pause:3.6\": failed to do request: Head \"https://registry-1.docker.io/v2/rancher/mirrored-pause/manifests/3.6\": dial tcp 44.205.64.79:443: i/o timeout"
```

**Reply by brandond** [reply]
I suspect that the rewrite was mangling the image name, and it was failing back to the default docker hub endpoint after failing to pull it from your mirror.
